### PR TITLE
Modules.t : Store module layout information

### PR DESCRIPTION
**Pull request by rgrinberg**
This is a part of #2305 extracted. It's not really clean yet, but it's ready for review. The 2nd part will completely replace Lib_modules. The reason why that part is harder is that it breaks the format of dune-package so it makes it very hard to test.

## Motivation for Modules.t

Currently, modules inside maps aren't cutting it. Here are some situation when we need more flexibility:

* Implementation of virtual modules have colliding names
* Wrapped modules need to be handled separately from other maps
* Qualified subdirs will require a the ability to store multiple module aliases
* The stdlib layout is quite complicated and is handled via a series of hacks spread around the codebase. It would be better if it was handled in one place.
* We'd like to transparently switch the compilation schemes for executables to use module aliases. Currently, that would be quite hard.
* The current codebase makes it too easy to construct partial dependency graphs. With `Modules.obj_map` one can be confident that the dependency graph will have an entry for every module.

## Possible TODOS

- [ ] I'd like to add another field to modules to distinguish their definition source: `Vlib | Impl`. With this, it will be easy to get rid of the `_no_vlib` suffixes and just have users check `Module.source_type` for example.
- [x] There's a couple of uses `Lib_modules` left. Those can be ported to `Modules.t` as well. A subsequent PR will remove `Lib_modules` completely.

### CODE REVIEW

**Comment by rgrinberg** [comment]
Compilation for all modules is now done in `Module_compilation.build_all`. This will make it very simple to switch to wrapped mode for executables.

File: src/modules.mli [hunk 1]
```diff
@@ -0,0 +1,51 @@
+open Stdune
```
Comment (diml) [inline src/modules.mli]: This module needs more documentation

**Comment by diml** [comment]
What's the relationship between `Lib_modules` and `Modules`?

**Comment by rgrinberg** [comment]
> What's the relationship between Lib_modules and Modules?

Lib_modules is the library specific predecessor to Modules. When I wanted to extend some of the functionality to executables as well, I realized that Lib_modules was not cutting it. Right now `Lib_modules` is kept around because it can be used to deserialize dune-package files. If I change the serialization format, I will need to recompile my switch to test things. So I will get rid of Lib_modules in the end, once I'm confident I won't need to run the tests.

**Comment by rgrinberg** [comment]
@diml I ended up getting rid of that module to build and generate alias files. I move the single function to `Module_compilation` and as a consequence, it no longer needs to be exposed in any public interface. Your comment is a good explanation however, so I kept it around in `Module_compilation.

**Comment by diml** [comment]
> Lib_modules is the library specific predecessor to Modules. When I wanted to extend some of the functionality to executables as well, I realized that Lib_modules was not cutting it. Right now Lib_modules is kept around because it can be used to deserialize dune-package files of already installed libraries. If I change the serialization format, I will need to recompile my switch to test things. So I will get rid of Lib_modules in the end, once I'm confident I won't need to run the tests.

If `Lib_modules` is only for the on-disk format, shouldn't it be just a separate decoder in `modules.ml`? 

> @diml I ended up getting rid of that module to build and generate alias files. I move the single function to Module_compilation and as a consequence, it no longer needs to be exposed in any public interface. Your comment is a good explanation however, so I kept it around in Module_compilation.

Sounds good.

**Comment by rgrinberg** [comment]
> If Lib_modules is only for the on-disk format, shouldn't it be just a separate decoder in modules.ml?

I will do that. Should just require a couple of tweaks. I was thinking that we can just break the format for dune-package, no? Since dune is no longer a build dependency.

@diml one thing that I've realized that is the `lib_interface_module` is a slightly confusing concept for virtual libraries. There are three different cases:

* It can exist in the virtual library only
* It can exist in the vlib and implementation if it's a virtual module.

And it can be confusing to for callers to handle either case. It seems that it's only being used for building the stdlib and for better errors for dependency cycles. The latter use case I've replace with `main_module_name` which is always well defined.

I think if we make stdlib first class in `Modules.t` we should just drop this concept altogether.

**Comment by diml** [comment]
> I will do that. Should just require a couple of tweaks. I was thinking that we can just break the format for dune-package, no? Since dune is no longer a build dependency.

Let's not break it for now. Maybe some non-opam users rely on it.

**Comment by diml** [comment]
More precisely, it's ok to use the new format when generating the file, but we should still be able to read the old format. So we don't need `Lib_modules.encode` and `Modules.to_lib_modules`. We should also hide `Lib_modules` inside `modules.ml` given that it's not used on its own.

**Comment by diml** [comment]
> And it can be confusing to for callers to handle either case. It seems that it's only being used for building the stdlib and for better errors for dependency cycles. The latter use case I've replace with main_module_name which is always well defined.

> I think if we make stdlib first class in Modules.t we should just drop this concept altogether.

That seems good to me. We should check that we can still build the compiler repository though if we touch this code.

File: test/blackbox-tests/test-cases/variants-external-declaration/run.t [hunk 2]
```diff
@@ -24,4 +24,4 @@ declared in the virtual library definition.
       (visibility public)
       (kind virtual)
       (intf)))
-    (wrapped true)))
+    (wrapped false)))
```
Comment (diml) [inline test/blackbox-tests/test-cases/variants-external-declaration/run.t]: Are that changes and the next one expected?

Comment (rgrinberg) [reply test/blackbox-tests/test-cases/variants-external-declaration/run.t]: Currently, it's a bit ambiguous what to set `wrapped` for libraries with only 1 module. I'll look into changing this.

**Comment by rgrinberg** [comment]
Okay, now stdlib compiles with minimal special casing. The primitives for qualified subdirs and wrapped executables are also in place. It is also now possible to  simplify the graph for implementation of vlib implementations by tweaking `Modules.obj_map` to also traverse the vlib modules. However, I'd like to think how to organize this better. It seems a bit weird to me that a module named `ocamldep` is running `ocamlobjinfo`. So I'd like something that sits on top of those.

**Comment by diml** [comment]
What about putting everything in `Dep_graph`?

**Comment by rgrinberg** [comment]
> What about putting everything in Dep_graph?

It will create cycles. Dep_graph was actually split to avoid cycles as quite a few places need to work withdependency graphs

**Comment by diml** [comment]
Ok. What about `Dep_graph_rules` then?

**Comment by rgrinberg** [comment]
Sounds good to me. Once we have qualified subdirs we can have something nice like:

```
ls dep_rules/
  ocamlobjinfo.ml
  ocamldep.ml
  dep_rules.ml # wraps both
```

File: src/module_compilation.ml [hunk 3]
```diff
@@ -159,27 +166,12 @@ let build_cm cctx ~dep_graphs ~precompiled_cmi ~cm_kind (m : Module.t) =
...
-                      | None -> flags
-                      | Some m' ->
-                        (* See comment in [Dune_file.Stdlib]. *)
-                        if Module.name m = Module.name m' then
-                          "-w" :: "-49" :: flags
-                        else
-                          "-open" :: Module.Name.to_string (Module.name m')
-                          :: flags)
+                  | Some _ ->
+                    (* XXX why aren't these just normal library flags? *)
```
Comment (diml) [inline src/module_compilation.ml]: Do you mean why we don't write `(flags -no-pervasives -nostdlib :standard)` in the `dune` file? It's just that we always need these flags for the stdlib, so it feels redundant to write them.

Comment (rgrinberg) [reply src/module_compilation.ml]: Okay. It just seems like if we can do with a more minimal feature set then we should do it. It's also easier to understand and change for users who aren't completely familiar with dune.

Comment (diml) [reply src/module_compilation.ml]: I guess. No objection to removing that, but the compiler `stdlib/dune`'s file needs to be adapted first.

File: test/blackbox-tests/test-cases/vlib/run.t [hunk 4]
```diff
@@ -354,20 +355,21 @@ Include variants and implementation information in dune-package
...
+     (alias_module
+      (name Vlib)
+      (obj_name vlib)
       (visibility public)
-      (kind virtual)
-      (intf)))
-    (wrapped true)))
+      (kind alias)
+      (impl))
+     (mode true))))
```
Comment (diml) [inline test/blackbox-tests/test-cases/vlib/run.t]: Actually, this looks odd. What does `(mode true)` mean?

Comment (diml) [reply test/blackbox-tests/test-cases/vlib/run.t]: I think we should keep the name `Wrapped` for the module and field name. Alternatively, we could rename the constructors as follow:

- `Simple true -> Wrapped`
- `Simple false -> Unwrapped`
- `Yes_with_transition x -> Wrapped_with_transition x`

Comment (rgrinberg) [reply test/blackbox-tests/test-cases/vlib/run.t]: Okay.

File: src/module_compilation.ml (hunk 3 above)
Comment (rgrinberg) [reply src/module_compilation.ml]: Hmm, I suppose this needs to be co-ordinated somehow. Make the PR to the compiler source after the release?

File: test/blackbox-tests/test-cases/vlib/run.t (hunk 4 above)
Comment (rgrinberg) [reply test/blackbox-tests/test-cases/vlib/run.t]: Actually, let's keep the renaming to another PR. It touches too much unrelated code.

I renamed the field to `wrapped` however.

**Comment by rgrinberg** [comment]
@diml i'm merging this because you already approved once and I fixed the one issue that you withdrew your approval for.

**Pull request merged**